src/modules/gamewisp/subscriber-emotes.js

```diff
@@ -0,0 +1,205 @@
+
+        emoteIDs.forEach(id => {
+            if (this.emotes.has(id)) {
+                emotes.push(this.emotes.get(id));
+            }
+        });
+
+        return emotes;
+    }
+
+    getEligibleEmote(code, user) {
+        if (settings.get('gwEmotes') === false) return;
+        if (!user || !this.emoteUsers) return;
+
+        const emoteIDs = this.emoteUsers.get(user.name);
+        if (!emoteIDs) return;
+
+        let emote;
+        for (const id of emoteIDs) {
+            emote = this.emotes.get(id);
+            if (emote && emote.code === code) {
+                return emote;
+            }
+        }
+
+        return;
+    }
+
+    joinRoom() {
+        const {name} = twitch.getCurrentChannel();
+
+        socketClient.joinRoom(name);
+    }
+
+    initializeRoom({userStore, emotes}) {
+        emotes = emotes.map(emoteData => buildEmote(emoteData)) || [];
+        emotes.forEach(emote => this.emotes.set(emote.id, emote));
+
+        for (const username in userStore) {
+            if (userStore.hasOwnProperty(username)) {
+                let userEmoteIDs = this.emotes.get(name);
+                if (!userEmoteIDs) {
+                    userEmoteIDs = [];
+                }
+
+                userEmoteIDs = [...userEmoteIDs, ...userStore[username]];
+
+                this.emoteUsers.set(username, userEmoteIDs);
+            }
+        }
+    }
+
+    updateRoom({emotes, user}) {
+        let newEmotes = emotes || [];
+
+        newEmotes = newEmotes.map(emoteData => buildEmote(emoteData));
+        newEmotes.forEach(emote => this.emotes.set(emote.id, emote));
+
+        let userEmoteIDs = this.emoteUsers.get(user.name);
+        if (!userEmoteIDs) {
+            userEmoteIDs = [];
+        }
+
+        userEmoteIDs = [...userEmoteIDs, ...user.emoteIDs];
+
+        this.emoteUsers.set(user.name, userEmoteIDs);
+    }
+
+    leaveRoom({user}) {
+        if (this.emoteUsers.get(user)) {
+            this.emoteUsers.delete(user);
+        }
+    }
+
+    removeEmote({emoteID}) {
+        const emoteUsers = [...this.emoteUsers.values()];
+
+        let idx;
+        emoteUsers.forEach(emoteIDs => {
+            idx = emoteIDs.indexOf(emoteID);
+
+            if (idx >= 0) {
+                emoteIDs.splice(idx, 1);
+            }
+        });
+
+        if (this.emotes.has(emoteID)) {
+            this.emotes.delete(emoteID);
+        }
+    }
+
+    addEmote({emote, emoteUsers}) {
+        emote = buildEmote(emote);
+
+        let userEmoteIDs;
+
+        emoteUsers.forEach(user => {
+            if (this.emoteUsers.has(user)) {
+                userEmoteIDs = this.emoteUsers.get(user);
+            } else {
+                userEmoteIDs = [];
+            }
+
+            userEmoteIDs = [...userEmoteIDs, emote.id];
+
+            this.emoteUsers.set(user, userEmoteIDs);
+            this.emotes.set(emote.id, emote);
+        });
+    }
+
+    newSubscriber({user, emotes, emoteIDs}) {
+        let userEmoteIDs;
```

Comment:
this can be simplified:

`let userEmoteIDs = this.emoteUsers.get(user) || [];`